File: app/src/main/java/com/davismiyashiro/expenses/datatypes/Expense.java
package com.davismiyashiro.expenses.datatypes;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.UUID;

/**
 * Object that encapsulates data for the Expense
 */
public class Expense implements Parcelable {
    private String mId;
    private String mDescription;
    private double mValue = 0;
    private String tabId;

    private void setId(String id) {
        mId = id;
    }

    public Expense(String description, double value, String tabid) {
        mId = UUID.randomUUID().toString();
        mDescription = description;
        mValue = value;
        tabId = tabid;
    }

    protected Expense(Parcel in) {
        mId = in.readString();
        mDescription = in.readString();
        mValue = in.readDouble();
        tabId = in.readString();
    }

    public static Expense retrieveExpense (String id, String description, double value, String tabId) {

        Expense expense = new Expense(description, value, tabId);
        expense.setId(id);
        return expense;
    }

    public static final Creator<Expense> CREATOR = new Creator<Expense>() {
        @Override
        public Expense createFromParcel(Parcel in) {
            return new Expense(in);
        }

        @Override
        public Expense[] newArray(int size) {
            return new Expense[size];
        }
    };

    public void setValue(double value) {
        mValue = value;
    }

    public void setDescription(String description) {
        mDescription = description;
    }

    public String getDescription() {

        return mDescription;
    }

    public double getValue() {
        return mValue;
    }

    public String getId() {
        return mId;
    }

    public String getTabId() {
        return tabId;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Expense expense = (Expense) o;

        if (Double.compare(expense.mValue, mValue) != 0) return false;
        if (mId != null ? !mId.equals(expense.mId) : expense.mId != null) return false;
        if (mDescription != null ? !mDescription.equals(expense.mDescription) : expense.mDescription != null)
            return false;
        return tabId != null ? tabId.equals(expense.tabId) : expense.tabId == null;

    }

    @Override
    public int hashCode() {
        int result;
        long temp;
        result = mId != null ? mId.hashCode() : 0;
        result = 31 * result + (mDescription != null ? mDescription.hashCode() : 0);
        temp = Double.doubleToLongBits(mValue);
        result = 31 * result + (int) (temp ^ (temp >>> 32));
        result = 31 * result + (tabId != null ? tabId.hashCode() : 0);
        return result;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(mId);
        dest.writeString(mDescription);
        dest.writeDouble(mValue);

        dest.writeString(tabId);
    }
}

File: app/src/main/java/com/davismiyashiro/expenses/datatypes/Expense.kt
/*
 * MIT License
 *
 * Copyright (c) 2019 Davis Miyashiro
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.davismiyashiro.expenses.datatypes

import android.os.Parcelable
import kotlinx.android.parcel.Parcelize

import java.util.UUID

/**
 * Object that encapsulates data for the Expense
 */
@Parcelize
data class Expense(var description: String, var value: Double, var tabId: String, var id: String) : Parcelable {
    constructor(description: String, value: Double, tabId: String) : this (description, value, tabId, UUID.randomUUID().toString())

    companion object {
        fun retrieveExpense(id: String, description: String, value: Double, tabId: String): Expense {

            val expense = Expense(description, value, tabId)
            expense.id = id
            return expense
        }
    }
}
